tikv/raftstore/peer.go

```diff
@@ -332,3 +401,599 @@ func NewPeer(storeId uint64, cfg *Config, engines *Engines, region *metapb.Regio
+		}
+		read.cmds = nil
+	}
+	p.pendingReads.reads = nil
+
+	for _, proposal := range p.applyProposals {
+		NotifyReqRegionRemoved(region.Id, proposal.Cb)
+	}
+	p.applyProposals = nil
+
+	log.Infof("%v destroy itself, takes %v", p.Tag, time.Now().Sub(start))
+	return nil
+}
+
+func (p *Peer) isInitialized() bool {
+	return p.peerStorage.isInitialized()
+}
+
+func (p *Peer) Region() *metapb.Region {
+	return p.peerStorage.Region()
+}
+
+/// Set the region of a peer.
+///
+/// This will update the region of the peer, caller must ensure the region
+/// has been preserved in a durable device.
+func (p *Peer) SetRegion(host *CoprocessorHost, region *metapb.Region) {
+	if p.Region().GetRegionEpoch().GetVersion() < region.GetRegionEpoch().GetVersion() {
+		// Epoch version changed, disable read on the localreader for this region.
+		p.leaderLease.ExpireRemoteLease()
+	}
+	p.Store().SetRegion(CloneRegion(region))
+
+	if !p.PendingRemove {
+		host.OnRegionChanged(p.Region(), RegionChangeEvent_Update, p.GetRole())
+	}
+}
+
+func (p *Peer) PeerId() uint64 {
+	return p.Peer.GetId()
+}
+
+func (p *Peer) GetRaftStatus() *raft.Status {
+	return p.RaftGroup.Status()
+}
+
+func (p *Peer) LeaderId() uint64 {
+	return p.RaftGroup.Raft.Lead
+}
+
+func (p *Peer) IsLeader() bool {
+	return p.RaftGroup.Raft.State == raft.StateLeader
+}
+
+func (p *Peer) GetRole() raft.StateType {
+	return p.RaftGroup.Raft.State
+}
+
+func (p *Peer) Store() *PeerStorage {
+	return p.peerStorage
+}
+
+func (p *Peer) IsApplyingSnapshot() bool {
+	return p.Store().IsApplyingSnapshot()
+}
+
+/// Returns `true` if the raft group has replicated a snapshot but not committed it yet.
+func (p *Peer) HasPendingSnapshot() bool {
+	return p.RaftGroup.GetSnap() != nil
+}
+
+func (p *Peer) Send(trans chan<- *rspb.RaftMessage, msgs []eraftpb.Message) error {
+	for _, msg := range msgs {
+		msgType := msg.MsgType
+		err := p.sendRaftMessage(msg, trans)
+		if err != nil {
+			return err
+		}
+		switch msgType {
+		case eraftpb.MessageType_MsgTimeoutNow:
+			// After a leader transfer procedure is triggered, the lease for
+			// the old leader may be expired earlier than usual, since a new leader
+			// may be elected and the old leader doesn't step down due to
+			// network partition from the new leader.
+			// For lease safety during leader transfer, transit `leader_lease`
+			// to suspect.
+			p.leaderLease.Suspect(time.Now())
+		default:
+		}
+	}
+	return nil
+}
+
+/// Steps the raft message.
+func (p *Peer) Step(m *eraftpb.Message) error {
+	if p.IsLeader() && m.GetFrom() != InvalidID {
+		p.PeerHeartbeats[m.GetFrom()] = time.Now()
+		// As the leader we know we are not missing.
+		p.leaderMissingTime = nil
+	} else if m.GetFrom() == p.LeaderId() {
+		// As another role know we're not missing.
+		p.leaderMissingTime = nil
+	}
+	return p.RaftGroup.Step(*m)
+}
+
+/// Checks and updates `peer_heartbeats` for the peer.
+func (p *Peer) CheckPeers() {
+	if !p.IsLeader() {
```

Comment:
We can check `len(p.PeerHeartbets) > 0` to avoid unneccessary memory allocation.

Reply:
Done